File: solutions/csharp/tree-building/1/TreeBuilding.cs
public class TreeBuildingRecord
{
    public int ParentId { get; set; }
    public int RecordId { get; set; }
}

public class Tree
{
    public int Id { get; set; }
    public int ParentId { get; set; }

    public List<Tree> Children { get; } = [];

    public bool IsLeaf => Children.Count == 0;
}

public static class TreeBuilder
{
    public static Tree BuildTree(IEnumerable<TreeBuildingRecord> records)
    {
        var orderedRecords = records.OrderBy(record => record.RecordId).ToArray();

        Validate(orderedRecords);

        var trees = CreateTrees(orderedRecords);

        LinkChildren(trees);

        return trees[0];
    }

    private static void Validate(TreeBuildingRecord[] records)
    {
        if (records.Length == 0)
            throw new ArgumentException();

        if (records[0].ParentId != 0)
            throw new ArgumentException();

        if (records.Select((record, id) => (record, id)).Any(IsInvalidRecord))
            throw new ArgumentException();
    }

    private static bool IsInvalidRecord((TreeBuildingRecord Record, int Id) pair)
    {
        var (record, id) = pair;

        return record.RecordId != id || (id != 0 && record.ParentId >= id);
    }

    private static Tree[] CreateTrees(TreeBuildingRecord[] records) =>
        records.Select(record => new Tree { Id = record.RecordId, ParentId = record.ParentId }).ToArray();

    private static void LinkChildren(Tree[] trees)
    {
        for (var id = 1; id < trees.Length; id++)
        {
            var tree = trees[id];
            trees[tree.ParentId].Children.Add(tree);
        }
    }
}

File: solutions/python/tree-building/1/tree_building.py
"""Build a binary tree from parent-child records."""

from dataclasses import dataclass, field


@dataclass(frozen=True)
class Record:
    """Describe a node and its parent by ID."""

    record_id: int
    parent_id: int


@dataclass
class Node:
    """Represent a tree node and its children."""

    node_id: int
    children: list["Node"] = field(default_factory=list)


def _validate_records(unordered_records):
    """Validate record IDs and parent relationships."""
    records = sorted(unordered_records, key=lambda record: record.record_id)

    record_ids = [record.record_id for record in records]

    if record_ids != list(range(len(records))):
        raise ValueError("Record id is invalid or out of order.")


    for record in records:
        if record.record_id == record.parent_id:
            if record.record_id != 0:
                raise ValueError("Only root should have equal record and parent id.")
        elif record.parent_id >= record.record_id:
            raise ValueError("Node parent_id should be smaller than its record_id.")

    return records


def BuildTree(records):  # pylint: disable=invalid-name
    """Build and return a tree from records, or None for no records."""
    if not records:
        return None

    ordered_records = _validate_records(records)

    nodes = {record.record_id: Node(record.record_id) for record in ordered_records}

    for record in ordered_records[1:]:
        nodes[record.parent_id].children.append(nodes[record.record_id])

    return nodes[0]
